### DISCUSSION TITLE
Port Forwarding With Tart

### DISCUSSION THREAD

**Discussion by raghav155** [cirruslabs/tart#855] [Q&A] [answered]
Hi, I'm trying to create a tart VM (which successfully gets created). I need to connect to a port in this VM from the host machine public IP. But it seems tart itself does not provide that functionality (port forwarding). Is there any work around anyone can suggest ?

**Comment by fkorotkov** [comment]
Tart itself doesn't have such functionality. You can try to use bridged network so VMs will get unique IPs or create a tunnel on the host with the default NAT network.

Alternatively you can check out [Orchard](https://github.com/cirruslabs/orchard) which automates and simplifies management of multiple hosts. Orchard [supports port forwarding](https://tart.run/blog/2023/04/28/ssh-over-grpc-or-how-orchard-simplifies-accessing-vms-in-private-networks/) out of the box.

**Reply by HenAmar** [reply]
Hey @fkorotkov, Hen from Harness. Are there any plans to support port forwarding capabilities directly, without requiring Orchestrator in the future? This feature is available on other macOS virtualization platforms and seems quite fundamental. While I understand that workarounds are possible, it feels like this is a basic feature that should be handled by the virtualization platform itself, rather than relying on additional components to manage it, especially when multiple VMs are being started and stopped.

**Reply by edigaryev** [reply]
Hi @HenAmar 👋 

>Are there any plans to support port forwarding capabilities directly, without requiring Orchestrator in the future?

Have you tried bridged networking in Tart?

You don't need an orchestrator like Orchard to be able to pass `--net-bridged` to `tart run`.

**Reply by HenAmar** [reply]
Hey @edigaryev 

I need to communicate with a service running inside a VM from the internet. To achieve this, there are two options:
	1.	Use bridge networking to assign the VM a dedicated IP.
	2.	Use port forwarding to route traffic to the VM.

However, since the bare-metal machine is in a farm with no DHCP server, assigning a dedicated IP to the VM via bridge networking isn’t feasible. Port forwarding is the only viable solution in this case.

**Reply by fkorotkov** [reply]
@HenAmar we'll take a look into this. We'll investigate adding this functionality into [Softnet](https://github.com/cirruslabs/softnet) which will allow you to have port forwarding as well as isolation unlike other solutions you mention.

**Reply by HenAmar** [reply]
@fkorotkov, I’m referring to something more akin to` hostfwd=tcp::3389-:3389` that we use in QEMU for port forwarding. Do you think this type of functionality could be implemented? If so, what would the timeframe be?

**Reply by fkorotkov** [reply]
I'll let @edigaryev comment on that since he's been actively looking into it. It seems we'll be able to cover this use case similarly to QEMU.

**Reply by HenAmar** [reply]
@edigaryev @fkorotkov Any updates on this? Could you provide a rough estimate of when it will be available?

**Reply by edigaryev** [reply]
@HenAmar this should be available once these two PRs (https://github.com/cirruslabs/tart/pull/990 and https://github.com/cirruslabs/softnet/pull/70) are merged 🙌

**Comment by edigaryev** [comment, marked as answer]
We've just released [Tart 2.23.0](https://github.com/cirruslabs/tart/releases/tag/2.23.0) and [Softnet 0.13.0](https://github.com/cirruslabs/softnet/releases/tag/0.13.0) and you can now expose VM's ports by passing the `--net-softnet-expose` option to `tart run`.

For example, to expose VM's 22 port on a TCP port 2222 on the host, use this command:

```
tart run --net-softnet-expose 2222:22 <VM>
```

If you're only interested in port forwarding and want to disable the Softnet restrictions (e.g. which prevent VM's communication with the local network), add a `--net-softnet-allow` option:

```
tart run --net-softnet-allow=0.0.0.0/0 --net-softnet-expose 2222:22 <VM>
```

Check out the `tart run --help` for more details.
